interface/forms/LBF/printable.php

```diff
@@ -142,111 +250,274 @@ function end_row()
-    if (strcmp($this_group, $last_group) != 0) {
-        end_group();
-        if (strlen($last_group) > 0) {
-            echo "<br />\n";
+    $this_group   = $frow['group_id'];
+    $titlecols    = $frow['titlecols'];
+    $datacols     = $frow['datacols'];
+    $data_type    = $frow['data_type'];
+    $field_id     = $frow['field_id'];
+    $list_id      = $frow['list_id'];
+    $edit_options = $frow['edit_options'];
+
+    $currvalue = '';
+    if ($formid || $visitid) {
+        $currvalue = lbf_current_value($frow, $formid, $visitid);
+    }
+
+    // Skip this field if skip conditions call for that.
+    // Note this also accumulates info for subsequent skip tests.
+    if (isSkipped($frow, $currvalue) == 'skip') continue;
+
+    if ($currvalue === FALSE) {
+        // Should not happen.
+        error_log("Function lbf_current_value() failed for field '$field_id'.");
+        continue;
+    }
+
+    // Skip this field if its do-not-print option is set.
+    if (strpos($edit_options, 'X') !== FALSE) continue;
+
+    $this_levels = $this_group;
+    $i = 0;
+    $mincount = min(strlen($this_levels), strlen($group_levels));
+    while ($i < $mincount && $this_levels[$i] == $group_levels[$i]) ++$i;
+    // $i is now the number of initial matching levels.
+
+    // If ending a group or starting a subgroup, terminate the current row and its table.
+    if ($group_table_active && ($i != strlen($group_levels) || $i != strlen($this_levels))) {
+        end_row();
+        echo " </table>\n";
+        $group_table_active = false;
+    }
+
+    // Close any groups that we are done with.
+    while (strlen($group_levels) > $i) {
+        $gname = $grparr[$group_levels]['grp_title'];
+        $group_levels = substr($group_levels, 0, -1); // remove last character
+        echo "</div>\n";
+        echo "</nobreak>\n";
+    }
+
+    // If there are any new groups, open them.
+    while ($i < strlen($this_levels)) {
+        end_row();
+        if ($group_table_active) {
+            echo " </table>\n";
+            $group_table_active = false;
         }
 
-        $group_name = substr($this_group, 1);
-        $last_group = $this_group;
-        echo "<b>" . text(xl_layout_label($group_name)) . "</b>\n";
-      
+        $group_levels .= $this_levels[$i++];
+        $gname = $grparr[substr($group_levels, 0, $i)]['grp_title'];
+        $subtitle = $grparr[substr($group_levels, 0, $i)]['grp_subtitle'];
+
+        // This is also for html2pdf. Telling it that the following stuff should
+        // start on a new page if there is not otherwise room for it on this page.
+        echo "<nobreak>\n";
+        echo "<div class='grpheader'>" . text(xl_layout_label($gname)) . "</div>\n";
         echo "<div class='section'>\n";
         echo " <table border='0' cellpadding='0'>\n";
+        echo "  <tr>";
+        for ($i = 1; $i <= $CPR; ++$i) {
+            $tmp = $i % 2 ? 'lcols1' : 'dcols1';
+            echo "<td class='$tmp'></td>";
```

Comment:
attr() around $tmp

Reply:
$tmp is just from a literal in the previous line, so it's clear there is no tainting from outside data.